synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,243 @@ def enumerate_leaves(node, depth):
                 yield m
 
 
+class _CacheEntry(Protocol):
+    """A protocol for cache entries used by `_ListNode` that allows it to cause
+    the cache entry to be dropped.
+    """
+
+    def drop_from_cache(self) -> None:
+        """Request the entry be dropped from the cache.
+
+        Note: This should call `.remove_from_list()` on all `_ListNodes`
+        referencing it.
+        """
+        ...
+
+
+P = TypeVar("P", bound=_CacheEntry)
```

Comment:
given that `_ListNode` doesn't call `drop_from_cache`, why is this bound necessary?

Reply:
It's called in `_expire_old_entries`, which iterates over the `_ListNode`?

Reply:
riiight, but doesn't `_expire_old_entries` know it's a `_ListNode[_CacheEntry]` ?

Reply:
Oh I see what you mean, yes you're right.

Reply:
Turns out we can get rid of `_CacheEntry` entirely.